src/main/java/org/broadinstitute/hellbender/utils/locusiterator/AlignmentContextIteratorBuilder.java

```diff
@@ -0,0 +1,178 @@
+     * @param isIncludeDeletions include reads with deletion on the loci in question
+     * @param isIncludeNs include reads with N on the loci in question
+     * @return iterator that produces AlignmentContexts ready for consumption (e.g. by a {@link org.broadinstitute.hellbender.engine.LocusWalker})
+     */
+    private static Iterator<AlignmentContext> createAlignmentContextIterator(final List<SimpleInterval> intervalsForTraversal,
+                                                                            final SAMFileHeader header,
+                                                                               final Iterator<GATKRead> readIterator,
+                                                                               final SAMSequenceDictionary dictionary,
+                                                                               final LIBSDownsamplingInfo downsamplingInfo,
+                                                                               final boolean isReference,
+                                                                               boolean emitEmptyLoci,
+                                                                               boolean isKeepUniqueReadListInLibs,
+                                                                               boolean isIncludeDeletions,
+                                                                               boolean isIncludeNs) {
+
+        // get the samples from the read groups
+        final Set<String> samples = header.getReadGroups().stream()
+                .map(SAMReadGroupRecord::getSample)
+                .collect(Collectors.toSet());
+
+        // get the LIBS
+        final LocusIteratorByState libs = new LocusIteratorByState(readIterator, downsamplingInfo, isKeepUniqueReadListInLibs, samples, header, isIncludeDeletions, isIncludeNs);
+        Iterator<AlignmentContext> iterator;
+
+        List<SimpleInterval> finalIntervals = intervalsForTraversal;
+        validateEmitEmptyLociParameters(emitEmptyLoci, dictionary, intervalsForTraversal, isReference);
+        if (emitEmptyLoci) {
+
+            // If no intervals were specified, then use the entire reference (or best available sequence dictionary).
+            if (!hasIntervals(finalIntervals)) {
+                finalIntervals = IntervalUtils.getAllIntervalsForReference(dictionary);
+            }
+            final IntervalLocusIterator intervalLocusIterator = new IntervalLocusIterator(finalIntervals.iterator());
+            iterator =  new IntervalAlignmentContextIterator(libs, intervalLocusIterator, header.getSequenceDictionary());
+
+        } else {
+            // prepare the iterator
+            iterator = hasIntervals(finalIntervals) ? new IntervalOverlappingIterator<>(libs, finalIntervals, header.getSequenceDictionary()) : libs;
```

Comment:
This will be clear if the hasIntervals condition is included in the if-else clause as following:

```java
if (emitEmptyLoci) {
    ...
} else if (hasIntervals(finalIntervals)) {
    ...
} else {
    iterator = libs;
}
```

Like that, it also can avoid the creation of `Iterator<AlignmentContext> iterator` by returning in each if-else branch the generated iterator...